### PR TITLE
Improve CI by combining workflows

### LINKED ISSUE #854: Improvements to the CI
- [ ] Only triggers when Rust or Cargo.toml/lock files are changed
- [ ] Investigate why crates are being recompiled even when cache has been recovered
- [ ] Merge check and test workflows ?

### PR DESCRIPTION
**Pull request by grtlr**
# Description of change

The goal of this PR is to fix some of the problems with our CI and hopefully also improve performance. It adds the following changes

* Only run CI on changes to the code (exclude `documentation/` and `**.md` files);
* Job-based system;
  * Combine `build` and `test` workflows into a single job;
  * Run other checks only if those have been successful;
 * More generous caching of dependencies and build artifacts;

## Links to any relevant issues

* closes #854 

## Type of change

- Bug fix (a non-breaking change which fixes an issue)
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

The caching is tricky to test, but I have seen significant speed ups. Also, the GitHub Actions UI is a bit cleaner now.

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ on:
       - stable
 
 jobs:
-  test:
+  setup:
```
Comment (thibault-martinez) [inline .github/workflows/ci.yml]: This ends up being the name shown in the interface so `setup` is not that great. You can either rename this or add a `name` property I think

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-name: Test
+name: CI
```
Comment (thibault-martinez) [inline .github/workflows/ci.yml:1]: Sorry I meant this one, `CI` is a too broad name for this IMO, all workflows are CI, maybe this one should be `Build/Test` or something ?

Comment (grtlr) [reply .github/workflows/ci.yml:1]: I think now that almost all jobs run in this workflow the name is appropriate, but I can also change it if you would prefer that.

**Review by pvdrz — DISMISSED** [review]

**Comment by grtlr** [comment]
Another nice effect is that we get better reporting within GitHub Action. Here is an example: https://github.com/iotaledger/bee/actions/runs/1503959223. We now can for example cancel almost the complete CI for a PR with a single click. Before we had to do this manually for each workflow.

**Review by Alex6323 — DISMISSED** [review]

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            ${{ matrix.os }}-${{ matrix.rust }}-
+            ${{ matrix.os }}-
+
+      - uses: actions-rs/clippy-check@v1
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          args: --all-targets --all-features -- -D warnings
+
+  format:
+    needs: build-and-test
```
Comment (thibault-martinez) [inline .github/workflows/ci.yml]: Is this really necessary ? I'd rather have this information as soon as possible.

Comment (grtlr) [reply .github/workflows/ci.yml]: Right now it first does a build. If that fails it gets reported immediately. To run the tests, you would first have to build  the individual crates anyway. A syntactic validation of testcases might take a littler longer, but in all other cases I think it shouldn't make a big difference time-wise. One downside is that the reporting is not as straight forward as it used to be (I think you have to click on the check to see which command went wrong).

But I could also be missing something. Why do you think the combined workflow would be slower?

Comment (thibault-martinez) [reply .github/workflows/ci.yml]: Just to be sure, I was speaking about the `format` check. It doesn't technically need build and test to perform. And I'm wondering if having to wait for them both just to discover that there is a formatting issue is wise or not.

Comment (grtlr) [reply .github/workflows/ci.yml]: Oh, sorry that was my mistake, I should get a coffee. I guess it is always a tradeoff: Spinning up machines takes time and prevents other jobs to run in the meantime. But I guess format runs fast enough to put it towards the front 👍 .

Comment (thibault-martinez) [reply .github/workflows/ci.yml]: All good, it runs in seconds when all the rest run in more than 15mins, I think it's fine to have it run first

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+      - dev
+      - stable
+    paths-ignore:
+      - 'documentation/**'
+      - '**.md'
+
+jobs:
+  build-and-test:
+    name: Build and Test
+    runs-on: ${{ matrix.os }}
```
Comment (thibault-martinez) [inline .github/workflows/ci.yml]: It triggers me a tiny bit that we use matrix before defining it :trollface:

Comment (grtlr) [reply .github/workflows/ci.yml]: Ah the beauty of declarative languages 😄

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+    - name: Cache cargo state
+      uses: actions/cache@v2
+      env:
+        cache-name: cache-cargo
+      with:
+        path: |
+          ~/.cargo/registry
+          ~/.cargo/git
+        key: ${{ matrix.os }}-${{ matrix.rust }}-build-${{ env.cache-name }}-${{ hashFiles('**/Cargo.toml') }}
```
Comment (thibault-martinez) [inline .github/workflows/ci.yml]: I think that `build` part is a relic from the previous version ? It should be covered by the cache name now.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+          ${{ matrix.os }}-${{ matrix.rust }}-build-${{ env.cache-name }}-
+          ${{ matrix.os }}-${{ matrix.rust }}-build-
+          ${{ matrix.os }}-${{ matrix.rust }}-
+          ${{ matrix.os }}-
+
+    - name: Build
+      uses: actions-rs/cargo@v1
+      with:
+        command: build
+        args: --all --release
```
Comment (thibault-martinez) [inline .github/workflows/ci.yml]: Wonder if there should be --all--features as well 🤔

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+            ubuntu-latest-nightly-
+            ubuntu-latest-
+
+      - uses: actions-rs/clippy-check@v1
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          args: --all-targets --all-features -- -D warnings
+
+  udeps:
+    needs: [build-and-test,clippy]
```
Comment (thibault-martinez) [inline .github/workflows/ci.yml]:
```suggestion
    needs: [build-and-test, clippy]
```

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+          command: install
+          args: cargo-udeps
+
+      - uses: actions-rs/cargo@v1
+        with:
+          command: udeps
+          args: --all-targets --all-features --backend=depinfo
+
+  licenses:
+    needs: [build-and-test,clippy]
```
Comment (thibault-martinez) [inline .github/workflows/ci.yml]:
```suggestion
    needs: [build-and-test, clippy]
```

File: .github/workflows/ci.yml (hunk 5 above)
Comment (grtlr) [reply .github/workflows/ci.yml]: We need it for more granularity in the caching. But it was in the wrong position here, so thanks for pointing that out!

**Review by Alex6323 — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+      - name: Cache cargo state
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-cargo
+        with:
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+            ~/.cargo/bin
+          key: ${{ matrix.os }}-${{ matrix.rust }}-${{ env.cache-name }}-${{ hashFiles('**/Cargo.toml') }}-build
```
Comment (thibault-martinez) [inline .github/workflows/ci.yml:57]: Why is this `-build` when it's later `-other` ?

Comment (grtlr) [reply .github/workflows/ci.yml:57]: We don't want the `clippy` job to overwrite our `build` cache (which was more expensive to create).

**Review by thibault-martinez — APPROVED** [review]

**Review by pvdrz — APPROVED** [review]

**Pull request merged**